### PR TITLE
Logit scale

### PR DESCRIPTION
**Pull request by iosonofabio**
As discussed earlier today on the mailing list, here the proposal of PR for a new scale, the LogitScale. This is essentially a nonlinear scale that is log both towards 0+ and log towards 1-. The exact transformation is:

x -> log10 (x / 1 - x)

(the 10 base of the log is just the most common case for plotting, but I could extend to natural or base 2 logs if there is interest). The inverse transform is the logistic function:

x -> 1 / (1 + 10^-x)

(again, 10 is just one possible base).

The LogitScale is useful when one has frequencies in a population (i.e. floats between 0 and 1) and both rare events and very common ones are interesting. For instance, say you ask about the fraction of people with blue eyes in various world populations, you want to spot even tiny deviations from zero or one.

Pierre Haessig had one implementation with example plots here:

http://nbviewer.ipython.org/gist/pierre-haessig/7e3e6a818edeb6819708

I had developed another, virtually identically one in the last few days, but this PR takes essentially his code and slightly modifies it for style and a few design points.

**Note**: there are a few FIXME in the code in case of doubts, please enlighten me on the best practice in those cases. Tests and docs also still missing.

### CODE REVIEW

**Comment by WeatherGod** [comment]
pings are the best way to get attention to possibly forgotten PRs.

**Comment by iosonofabio** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

pings?

sry I am not sure what that is.

Thx
F

On 02/16/2015 03:53 PM, Benjamin Root wrote:
> pings are the best way to get attention to possibly forgotten PRs.
> 
> — Reply to this email directly or view it on GitHub 
> <https://github.com/matplotlib/matplotlib/pull/3753#issuecomment-74519376>.
>
-----BEGIN PGP SIGNATURE-----
Version: GnuPG v2

iQEcBAEBCAAGBQJU4gkFAAoJELvfSqOdeQF5f88H/0ml/8/6mMYJCof/JEM+drjO
ypHd4MwkRL6WKkUY/XaOcsiAm1KO6WHdDBhkPmI0ySpFUp+m7kno8gN9yr1Ga9p2
6qv6jjfx+i+G1sKsgsZHeJgXduEt5X5aOz5eXJonzxwjxRzxYylMrb+8A9z4jY6B
kUsvUn5VJNYhXdMNs/LSryfrrIM6PopkyzthXN7KdXbwROVbN+yvqlX6xfqBxKqG
IFw1pwFJ1Nsei1nJf6G/AQneGg2wxuqRbM2QmAuHVAn49miT9BV+aFTAjPXzZbV0
nk2WHs1Z1fvZ9lIAY+fqBKgJSQkEVuNMmJMICRpnQf7+QoIFLuxB2+DhC+Gxfjk=
=W5nS
-----END PGP SIGNATURE-----

**Comment by WeatherGod** [comment]
What you did just now was a "ping". The developers get email notifications
whenever a comment is added to an Issue or PR. We are very much like a
bunch of cats. Difficult to herd, and tend to lose focus easily. Pinging
helps get our attention to the issues that needs it most.

Also notice that your issue was milestoned for the version 1.5.0 release.
We have just now released version 1.4.3, so we can now focus on new
features for version 1.5. We don't follow a rapid release schedule like
some projects, so it can seem like a very long time to accept new features.
This can be frustrating for some, and we apologize for that, but we have
not forgotten you, and we do greatly appreciate your contribution!

**Comment by tacaswell** [comment]
It also works best if you include a @iosonofabio (@ + username) to make sure the dev involved gets an email.

**Comment by iosonofabio** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

Hi,

Ah ok thanks for the explanation. Sorry if the other email sounded
nasty, it was not meant to.

I am not in a hurry to see it released, 1.5 is alright, I am just a
bit wary because the more time is passing, the less well I remember
the codebase, so making changes becomes harder for me. That's why I'd
like to see it accepted (in the dev version) sometime soon, assuming
everything's ok with it.

Cheers, thx again!
Fabio

On 02/16/2015 05:07 PM, Benjamin Root wrote:
> What you did just now was a "ping". The developers get email
> notifications whenever a comment is added to an Issue or PR. We are
> very much like a bunch of cats. Difficult to herd, and tend to lose
> focus easily. Pinging helps get our attention to the issues that
> needs it most.
> 
> Also notice that your issue was milestoned for the version 1.5.0
> release. We have just now released version 1.4.3, so we can now
> focus on new features for version 1.5. We don't follow a rapid
> release schedule like some projects, so it can seem like a very
> long time to accept new features. This can be frustrating for some,
> and we apologize for that, but we have not forgotten you, and we do
> greatly appreciate your contribution!
> 
> On Mon, Feb 16, 2015 at 10:13 AM, Fabio Zanini
> <notifications@github.com> wrote:
> 
> pings?
> 
> sry I am not sure what that is.
> 
> Thx F
> 
> On 02/16/2015 03:53 PM, Benjamin Root wrote:
>> pings are the best way to get attention to possibly forgotten
>> PRs.
> 
>> — Reply to this email directly or view it on GitHub <
> https://github.com/matplotlib/matplotlib/pull/3753#issuecomment-74519376>.
>
> 
>> 
>> — Reply to this email directly or view it on GitHub 
>> <https://github.com/matplotlib/matplotlib/pull/3753#issuecomment-74522439>
>>
>> 
.
>> 
> 
> — Reply to this email directly or view it on GitHub 
> <https://github.com/matplotlib/matplotlib/pull/3753#issuecomment-74531226>.
>
-----BEGIN PGP SIGNATURE-----
Version: GnuPG v2

iQEcBAEBCAAGBQJU4hgQAAoJELvfSqOdeQF5ne8H/A5BHY18bxdQGInb6worhvrU
pBauDLCmXW0eQ3pUV6NUsbGelB+3xEQygt2O3FnzToZJCKlY7bNEPg+vcYsfW+Yc
z+ogmOnu6CUH+Ug1ZHNMKFh+ykEqZIJMQdgN1i/OAQFFwU5GWkehqq/08PBYMTbZ
rqSQso68WopHaYHUZ5oAHWDVrM4PjL7awJNESmRCIcjCY9DRRu/a2y8qLRx8IZ8J
WZkaT62cdLMJwz4yRoBHDcGAHoKBlt+YzmVGWVmUY/JDu6aJM7w3DCYQZZ3YTH7E
rt60kMjjNCj/HCbcgfWb83WaIz14YG9L0+EAiIW6Nm7z7JXjUpVG9FibgtpC5RY=
=mOmO
-----END PGP SIGNATURE-----

File: lib/matplotlib/scale.py [hunk 1]
```diff
@@ -478,10 +478,110 @@ def get_transform(self):
...
+    masked. If all values are inside ]0, 1[, the original array is
+    returned.
+    """
+    mask = (a <= 0.0) | (a >= 1.0)
+    if mask.any():
+        return ma.MaskedArray(a, mask=mask)
+    return a
+
+
+def _clip_non_logit(a):
```
Comment (tacaswell) [inline lib/matplotlib/scale.py]: You should make a copy of the input array either here or where `_handle_non_affine` is called.  Modifying user objects is bad practice.

File: lib/matplotlib/scale.py [hunk 2]
```diff
@@ -478,10 +478,110 @@ def get_transform(self):
...
 
+def _mask_non_logit(a):
+    """
+    Return a Numpy masked array where all values outside ]0, 1[ are
+    masked. If all values are inside ]0, 1[, the original array is
+    returned.
+    """
+    mask = (a <= 0.0) | (a >= 1.0)
+    if mask.any():
+        return ma.MaskedArray(a, mask=mask)
```
Comment (tacaswell) [inline lib/matplotlib/scale.py]: I would just do `a[mask] = np.nan` .  I think we are trying to move away from using masked arrays internally due to the issues with them.  Again, make sure a copy gets made before mutating a user object.

**Comment by tacaswell** [comment]
Could you rebase (sorry, one more time) to remove the pdf and svg test images (unless there is a compelling reason that this test should be backend dependent)?  Adding ` extensions=['png'],` to the `@image_comparison` decorator will make it only do the png testing.  The repo is already very large and limiting the growth of test images is one of the ways we are trying to keep it from getting too much worse.

**Comment by iosonofabio** [comment]
Both suggestions followed. Now I'll rebase.

**Comment by iosonofabio** [comment]
ping

File: lib/matplotlib/scale.py [hunk 3]
```diff
@@ -478,10 +478,111 @@ def get_transform(self):
         return self._transform
 
 
+def _mask_non_logit(a):
+    """
+    Return a Numpy masked array where all values outside ]0, 1[ are
```
Comment (efiring) [inline lib/matplotlib/scale.py]: This docstring does not match what the function is doing.  It is always returning a copy of the input array in which any values outside ]0, 1[ are replaced by nan.

File: lib/matplotlib/scale.py [hunk 4]
```diff
@@ -478,10 +478,111 @@ def get_transform(self):
...
+    masked. If all values are inside ]0, 1[, the original array is
+    returned.
+    """
+    a = a.copy()
+    mask = (a <= 0.0) | (a >= 1.0)
+    a[mask] = np.nan
+    return a
+
+
+def _clip_non_logit(a):
```
Comment (efiring) [inline lib/matplotlib/scale.py]: I think you could achieve what you want in one line with `return np.clip(a, 1e-300, 1 - 1e-300)`.  Either way, though, you are assuming `a` is double precision.  Is this a safe assumption?

File: lib/matplotlib/scale.py [hunk 5]
```diff
@@ -478,10 +478,111 @@ def get_transform(self):
...
+        if nonpos == 'mask':
+            self._handle_nonpos = _mask_non_logit
+        else:
+            self._handle_nonpos = _clip_non_logit
+        self._nonpos = nonpos
+
+    def transform_non_affine(self, a):
+        """logit transform (base 10), masked or clipped"""
+        a = self._handle_nonpos(a)
+        if isinstance(a, ma.MaskedArray):
```
Comment (efiring) [inline lib/matplotlib/scale.py]: It can't be a masked array here.

File: lib/matplotlib/scale.py [hunk 6]
```diff
@@ -478,10 +478,111 @@ def get_transform(self):
         return self._transform
 
 
+def _mask_non_logit(a):
+    """
+    Return a Numpy masked array where all values outside ]0, 1[ are
+    masked. If all values are inside ]0, 1[, the original array is
+    returned.
+    """
+    a = a.copy()
```
Comment (efiring) [inline lib/matplotlib/scale.py]: It looks like you switched in midstream from using masked arrays to using nans.  Does this need to handle masked array inputs?  If so, and if you can subsequently use only nans for bad values, you can do this:
```python
a = np.ma.masked_invalid(a).filled(np.nan)
```
Regardless of input, that will leave you with an ndarray with nothing but valid numbers and nans.  It handles inf as well as nan.

File: lib/matplotlib/scale.py [hunk 7]
```diff
@@ -478,10 +478,111 @@ def get_transform(self):
...
+
+    def set_default_locators_and_formatters(self, axis):
+        # ..., 0.01, 0.1, 0.5, 0.9, 0.99, ...
+        axis.set_major_locator(LogitLocator())
+        axis.set_major_formatter(LogitFormatter())
+        axis.set_minor_locator(LogitLocator(minor=True))
+        axis.set_minor_formatter(LogitFormatter())
+
+    def limit_range_for_scale(self, vmin, vmax, minpos):
+        return (vmin <= 0 and minpos or vmin,
```
Comment (efiring) [inline lib/matplotlib/scale.py]: I find it hard to parse this expression; and it needs a docstring at least to explain what `minpos` is.

File: lib/matplotlib/tests/test_scale.py [hunk 8]
```diff
@@ -14,6 +14,20 @@ def test_log_scales():
     ax.axhline(24.1)
 
 
+@image_comparison(baseline_images=['logit_scales'], remove_text=True,
+                  extensions=['png'])
+def test_logit_scales():
+    ax = plt.subplot(111, xscale='logit')
+
+    # Typical exctinction curve for logit
```
Comment (efiring) [inline lib/matplotlib/tests/test_scale.py]: Typo: -> extinction

File: lib/matplotlib/ticker.py [hunk 9]
```diff
@@ -807,6 +807,26 @@ def __call__(self, x, pos=None):
                                                          nearest_long(fx))
 
 
+class LogitFormatter(Formatter):
+    '''Probability formatter (using Math text)'''
+    def __call__(self, x, pos=None):
+        s = ''
+        if 0.01 <= x <= 0.99:
+            if x in [.01, 0.1, 0.5, 0.9, 0.99]:
```
Comment (efiring) [inline lib/matplotlib/ticker.py]: Does this have to be so restrictive?  Only those few values?

**Comment by iosonofabio** [comment]
He Eric, thanks for all the comments, which I will address ASAP.
I must confess: if those came out *before* I rebase for the n-th time, I'd not particularly mind. But as they are legitimate, no problem.

File: lib/matplotlib/scale.py (hunk 4 above)
Comment (iosonofabio) [inline lib/matplotlib/scale.py]: just copying from _mask_non_positives, to keep the codebase tidy.

File: lib/matplotlib/scale.py (hunk 6 above)
Comment (iosonofabio) [inline lib/matplotlib/scale.py]: Ok, so I think keeping the logit _mask and _clip functions analogous to the log ones (_mask_non_positives etc) is good for the codebase, because they have the exact same funcion.

But the *_non_positives functions have bugs, i.e. the docstring is wrong (it mentions masked arrays) and the _clip function modifies user input. So I'll correct the _non_positive function as well, I am against a PR with the logit/log parts of the code doing different things.

File: lib/matplotlib/scale.py (hunk 5 above)
Comment (iosonofabio) [inline lib/matplotlib/scale.py]: see below: the code for the LogTransforms has the same problem all over the place, I was just copying from there to keep the codebase uniform. But now if mpl is moving away from masked arrays for whatever reason, I'll just correct all the LogTransforms classes too.

File: lib/matplotlib/scale.py (hunk 7 above)
Comment (iosonofabio) [inline lib/matplotlib/scale.py]: I'll add a short docstring, but it's the same like the analog function for LogTransforms. Again, I'm just being consistent with the codebase.

File: lib/matplotlib/tests/test_scale.py (hunk 8 above)
Comment (iosonofabio) [inline lib/matplotlib/tests/test_scale.py]: fixed

File: lib/matplotlib/ticker.py (hunk 9 above)
Comment (iosonofabio) [inline lib/matplotlib/ticker.py]: fair enough, fixed

**Comment by efiring** [comment]
The ticker module needs more work, such as #4172, so any remaining fine-tuning of the logit functionality can be done later.

**Comment by iosonofabio** [comment]
Cool, thanks!

**Comment by pierre-haessig** [comment]
Hi @iosonofabio, great to this that this PR went in! I just have a question, is the logit scale also usable for the `norm` argument of `plt.imshow` ? (like http://stackoverflow.com/a/2546622) ? 

Last week I was plotting the transition matrix of a Markov chain, with many values close to 0 and many other values close to 1. Plotting the logit transform of that matrix with a diverging colormap really gave a nice result.

**Comment by WeatherGod** [comment]
Currently, I don't think that is possible, but there have been discussion
in past years to merge the scales and norm concepts in some manner. It
isn't something that can be represented via a single codebase, I don't
think, but it would be interesting to see if we could create a Normalize
subclass that could accept a scale.

**Comment by pierre-haessig** [comment]
To illustrate this notion of logit norm, I've created a Notebook example http://nbviewer.ipython.org/gist/pierre-haessig/2d4b46078d46e6fc3842

Thanks Ben for your feedback. I guess that the easiest path would be to create a new [Normalize](http://matplotlib.org/api/colors_api.html#matplotlib.colors.Normalize) subclass and leave for fancy refactoring for another day. And it's not that clear that this would lead to that much code duplication.

**Comment by tacaswell** [comment]
+1 on doing the simplest thing that will solve this problem.

**Comment by iosonofabio** [comment]
I am unfortunately terribly busy these days but I would be interested in
seeing that implemented, simplest thing is the best for now I think.
